firedrake/preconditioners/pmg.py

```diff
@@ -95,124 +128,147 @@ def initialize(self, pc):
+        add_hook(parent, setup=partial(push_parent, pdm, parent), teardown=partial(pop_parent, pdm, parent), call_setup=True)
+        add_hook(parent, setup=partial(push_appctx, pdm, ctx), teardown=partial(pop_appctx, pdm, ctx), call_setup=True)
 
-        ppc.setFromOptions()
-        ppc.setUp()
-        self.ppc = ppc
-
-    def apply(self, pc, x, y):
-        return self.ppc.apply(x, y)
-
-    def applyTranspose(self, pc, x, y):
-        return self.ppc.applyTranspose(x, y)
+        self.ppc = self.configure_pmg(pc, pdm)
+        self.ppc.setFromOptions()
+        self.ppc.setUp()
 
     def update(self, pc):
         pass
 
     def coarsen(self, fdm, comm):
+        # Coarsen the _SNESContext of a DM fdm
+        # return the coarse DM cdm of the coarse _SNESContext
         fctx = get_appctx(fdm)
+
+        # Have we already done this?
+        cctx = fctx._coarse
+        if cctx is not None:
+            return cctx.J.arguments()[0].function_space().dm
+
+        parent = get_parent(fdm)
+        assert parent is not None
+
         test, trial = fctx.J.arguments()
         fV = test.function_space()
-        fu = fctx._problem.u
-
         cele = self.coarsen_element(fV.ufl_element())
         cV = firedrake.FunctionSpace(fV.mesh(), cele)
         cdm = cV.dm
 
+        def get_max_degree(ele):
+            if isinstance(ele, MixedElement):
+                return max(get_max_degree(sub) for sub in ele.sub_elements())
+            else:
+                N = ele.degree()
+                try:
+                    N = max(N)
+                except TypeError:
+                    pass
+                return N
+
+        def coarsen_quadrature(df, Nf, Nc):
+            # Coarsen the quadrature degree in a dictionary
+            # such that the ratio of quadrature nodes to interpolation nodes (Nq+1)/(Nf+1) is preserved
+            if isinstance(df, dict):
+                Nq = df.get("quadrature_degree", None)
+                if Nq is not None:
+                    dc = dict(df)
+                    dc["quadrature_degree"] = max(2*Nc+1, ((Nq+1) * (Nc+1) + Nf) // (Nf+1) - 1)
+                    return dc
+            return df
+
+        def coarsen_form(form, Nf, Nc, replace_d):
+            # Coarsen a form, by replacing the solution, test and trial functions, and
+            # reconstructing each integral with a corsened quadrature degree.
+            # If form is not a Form, then return form.
+            return Form([f.reconstruct(metadata=coarsen_quadrature(f.metadata(), Nf, Nc))
+                         for f in replace(form, replace_d).integrals()]) if isinstance(form, Form) else form
+
+        def coarsen_bcs(fbcs):
+            cbcs = []
+            for bc in fbcs:
+                cV_ = cV
+                for index in bc._indices:
+                    cV_ = cV_.sub(index)
+                cbc_value = self.coarsen_bc_value(bc, cV_)
+                cbcs.append(firedrake.DirichletBC(cV_, cbc_value,
```

Comment:
`type(bc)(cV_, ...)`

I think maybe a bigger change (open an issue to do it) is to provide `reconstruct` methods for BC objects so that you don't have to dispatch based on type when doing this stuff, but not sure.